File: Neetcode - ALL/Arrays and Hashing/top-k-frequent-elements.cpp

// https://leetcode.com/problems/top-k-frequent-elements/

bool comparePairs(pair<int, int>a, pair<int, int>b){
    return a.first < b.first;
}

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        
        unordered_map<int, int> mp;
        for(auto num : nums) mp[num]++;

        priority_queue< pair<int, int>, vector<pair<int, int>>, decltype(&comparePairs)> pq(comparePairs);
        for(auto it : mp){
            pq.push({it.second, it.first});
        }

        vector<int> ans;
        while(k--){
            ans.push_back(pq.top().second);
            pq.pop();
        }

        return ans;
    }
};

File: Neetcode Questions in Js/Arrays and Hashing/top-k-frequent-elements.js

// https://leetcode.com/problems/top-k-frequent-elements/

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    const freq = {};
    nums.forEach((num) => {
        if(freq[num]) freq[num].count++;
        else freq[num] = {num, count : 1}
    })
    let temp = [];
    for(const key in freq) temp.push(freq[key]);
    temp = temp.sort((a, b) => a.count - b.count);
    // console.log(temp);
    const result = [];
    while(k--){
        result.push(temp.pop().num)
    }
    return result;
};